feat(appchains): Introduce module + register AVS

MaxMustermann2 opened the pull request:
This PR makes partial progress towards app chains as an AVS. It includes the capability to accept transactions (signed by anyone) to create a new chain-based AVS to activate at the end of the current epoch. The genesis state of the app chain is made available by the coordinator module for importing by the subscriber module of the appchain. 

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Enhanced documentation for clarity regarding the slashing keeper's role.
	- Added a method for validating asset IDs, improving error handling and data integrity in asset management.
	- Introduced gRPC query methods for retrieving application chain parameters and subscriber genesis data.
	- Added functionality for managing subscriber chains, including registration and storage of genesis states.

- **Bug Fixes**
	- Improved accuracy of unbonding time estimation.

- **Documentation**
	- Expanded comments and documentation across various files to assist future developers.

- **Chores**
	- Established foundational structures for error handling, event management, and codec registration within the appchain.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

MaxMustermann2 committed:
feat(appchain): scaffold coordinator

MaxMustermann2 committed:
feat(appchain): scaffold subscriber

MaxMustermann2 committed:
feat(epochs): create number + id structure

MaxMustermann2 committed:
feat(appchain): add coordinator params

MaxMustermann2 committed:
fix(epochs): invalidate negative duration

MaxMustermann2 committed:
fix(avs): stateful validation of staking assets

MaxMustermann2 committed:
refactor(dogfood): carve `sortByPower` out

MaxMustermann2 committed:
fix(dogfood): avoid duplicate validation

x/avs handles the assetID validation

MaxMustermann2 committed:
fix(dogfood): replace with max unbonding duration

MaxMustermann2 committed:
doc(app): add comment about slashing frac storage

MaxMustermann2 committed:
feat(appchain): allow sub chain creation

- Anyone can create a subchain via a simple transaction. The genesis
  file is available at the end of the current epoch.
- If no operators opt in by that time, it is dropped.
- If the state is available, it is queried via CLI and added to the
  genesis file of the subchain.
- The genesis file, with data from other modules, starts the chain once
  enough vote power is live

bwhour commented on x/appchain/common/types/events.go:
```diff
@@ -0,0 +1,16 @@
+	// EventTypeTimeout       = "timeout"
+	AttributeKeyAckSuccess    = "success"
+	AttributeKeyAck           = "acknowledgement"
+	AttributeKeyAckError      = "error"
```
why not use ack_error, it's easiler to understand ?

bwhour commented on x/appchain/common/types/events.go:
```diff
@@ -0,0 +1,16 @@
+package types
+
+const (
+	// EventTypeTimeout       = "timeout"
```
remove this line.

MaxMustermann2 commented on proto/exocore/appchain/coordinator/v1/params.proto:
```diff
@@ -0,0 +1,28 @@
+
+option go_package = "github.com/ExocoreNetwork/exocore/x/appchain/coordinator/types";
+
+// Params is the parameters for the appchain coordinator module.
+message Params {
+    // template_client is the IBC template client.
+    ibc.lightclients.tendermint.v1.ClientState template_client = 1;
+    // trusting_period_fraction is the multiplier applied on the subscriber's unbonding duration to determine the IBC trusting period.
+    string trusting_period_fraction = 2;
+    // ibc_timeout_period is the timeout period for IBC packets. While our system is largely created with epochs as a unit of time (and not standard durations), this is an exception since it is used directly by the IBC codebase.
+    google.protobuf.Duration ibc_timeout_period = 3
+    [ (gogoproto.nullable) = false, (gogoproto.stdduration) = true, (gogoproto.customname) = "IBCTimeoutPeriod" ];
+    // init_timeout_period is the period within which the subscriber chain must make a connection with the coordinator, after being spawned.
+    exocore.epochs.v1.Epoch init_timeout_period = 4
+    [ (gogoproto.nullable) = false ];
+    // vsc_timeout_period is the period within which the subscriber chain must respond to a VSC request, after it is sent.
+    exocore.epochs.v1.Epoch vsc_timeout_period = 5
+    [ (gogoproto.nullable) = false, (gogoproto.customname) = "VSCTimeoutPeriod" ];
+    // TODO: slashing and reward related parameters
+}
```
Yes

MaxMustermann2 replied:
Actually I am removing this TODO comment since these are params of the subscriber and not the coordinator.

MaxMustermann2 commented on x/appchain/coordinator/keeper/register.go:
```diff
@@ -0,0 +1,91 @@
+	}
+	// this value is required by the AVS module when making edits or deleting the AVS. as always, we round it up for the
+	// current epoch. it can never be 0, because both the durations are positive.
+	unbondingEpochs := math.Ceil(float64(req.SubscriberParams.UnbondingPeriod) / float64(epochInfo.Duration))
```
Give an example of how this might happen.

MaxMustermann2 commented on proto/exocore/appchain/common/v1/common.proto:
```diff
@@ -0,0 +1,104 @@
+// and each of the subscriber modules. These types (or parts thereof) are stored
+// within the module states but not sent over the wire.
+
+// Params defines the parameters for the subscriber module. TODO(mm): These must be deny listed
+// for edits via governance on the subscriber chain to prevent the subcriber participants
+// from unilaterally alterting parameters (like reward %) for their own benefit.
```
Create an issue.

MaxMustermann2 committed:
chore(lint)

MaxMustermann2 committed:
Merge `develop`

MaxMustermann2 committed:
fix(test): make avs test pass

The AVS module now validates the assetIDs passed to it

MaxMustermann2 committed:
fix(appchain): remove superfluous test

MaxMustermann2 committed:
fix(epochs): update expected test message

negative duration is also not supported

MaxMustermann2 committed:
chore(lint)

MaxMustermann2 committed:
chore(lint): proto lint

MaxMustermann2 committed:
chore: run proto gen

MaxMustermann2 committed:
fix(appchain): remove floating point arithmetic

MaxMustermann2 committed:
fix(appchain): remove wrong comment

MaxMustermann2 commented on x/appchain/subscriber/keeper/genesis.go:
```diff
@@ -0,0 +1,18 @@
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+func (k Keeper) InitGenesis(ctx sdk.Context, gs types.GenesisState) []abci.ValidatorUpdate {
+	k.SetParams(ctx, gs.Params)
+	return []abci.ValidatorUpdate{}
```
It is a key value store. `SetParams` cannot error.

MaxMustermann2 commented on x/appchain/coordinator/keeper/genesis.go:
```diff
@@ -0,0 +1,18 @@
+	return []abci.ValidatorUpdate{}
+}
+
+func (k Keeper) ExportGenesis(ctx sdk.Context) *types.GenesisState {
+	return &types.GenesisState{
+		Params: k.GetParams(ctx),
+	}
```
`GetParams` cannot error.

MaxMustermann2 commented on x/appchain/coordinator/keeper/genesis.go:
```diff
@@ -0,0 +1,18 @@
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+func (k Keeper) InitGenesis(ctx sdk.Context, gs types.GenesisState) []abci.ValidatorUpdate {
+	k.SetParams(ctx, gs.Params)
+	return []abci.ValidatorUpdate{}
```
`SetParams` cannot error.

MaxMustermann2 commented on x/appchain/coordinator/types/genesis.go:
```diff
@@ -0,0 +1,22 @@
+	return &GenesisState{Params: params}
+}
+
+// Validate performs basic genesis state validation returning an error upon any
+// failure.
+func (gs GenesisState) Validate() error {
+	if err := gs.Params.Validate(); err != nil {
+		return err
+	}
+	// TODO: validate anything else added here
+	return nil
```
When new data is added I will add the corresponding validation myself. There is no data right now to validate.

MaxMustermann2 commented on x/appchain/subscriber/keeper/params.go:
```diff
@@ -0,0 +1,22 @@
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// SetParams sets the appchain coordinator parameters.
+func (k Keeper) SetParams(ctx sdk.Context, params commontypes.SubscriberParams) {
+	store := ctx.KVStore(k.storeKey)
+	bz := k.cdc.MustMarshal(&params)
+	store.Set(types.ParamsKey(), bz)
+}
```
Added comment on the function to make it the responsibility of the caller.

MaxMustermann2 committed:
doc(appchain): add comment about params expect

MaxMustermann2 commented on proto/exocore/appchain/subscriber/v1/query.proto:
```diff
@@ -0,0 +1,28 @@
+
+package exocore.appchain.subscriber.v1;
+
+import "exocore/appchain/common/v1/common.proto";
```
The file exists within the `proto` subfolder.

MaxMustermann2 committed:
fix(appchain): remove superfluous fmt

MaxMustermann2 committed:
feat(appchain): report invalid json

MaxMustermann2 committed:
fix(utils): remove SliceStable and use Slice

MaxMustermann2 commented on proto/exocore/appchain/common/v1/common.proto:
```diff
@@ -0,0 +1,104 @@
+
+    // These are params related to the slashing module. Requests are received
+    // from the subscriber and slashed according to these params. Since signing
+    // can only be tracked by the subscriber chain, we do not have any parameters
+    // here that can be used to configure the signing window and the number of
+    // blocks that should be signed in it. Conversely, the subscriber chain
+    // does not do anything with these parameters (even though they are shared)
+    // since slashing is done by the coordinator chain.
+    // Operators should refer to the genesis file of the subscriber chain to
+    // check their comfort with these values before onboarding the chain.
+
+    // slash_fraction_downtime is the fraction of the stake that is slashed when a validator is found to be offline.
+    string slash_fraction_downtime = 15;
+    // downtime_jail_duration is the duration of the jail period for a validator
+    // after they have been found to be offline for too long.
+    google.protobuf.Duration downtime_jail_duration = 16
+      [(gogoproto.nullable) = false, (amino.dont_omitempty) = true, (gogoproto.stdduration) = true];
+    // slash_fraction_double_sign is the fraction of the stake that is slashed when a validator is found to have double signed.
+    string slash_fraction_double_sign = 17;
+}
+
+// SubscriberGenesisState is the genesis state of a subscriber at the time of
+// it being provisioned by Exocore, as stored in the coordinator module.
+message SubscriberGenesisState {
+  // params is the parameters of the subscriber module, as generated.
+  SubscriberParams params = 1 [(gogoproto.nullable) = false];
+  // coordinator is the coordinator information for the subscriber.
+  CoordinatorInfo coordinator = 2 [ (gogoproto.nullable) = false ];
+}
+
+// CoordinatorInfo is the information about the coordinator chain that is stored within the subscriber chain's subscriber module.
+message CoordinatorInfo {
+  // client_state is the client state of the coordinator chain.
+  ibc.lightclients.tendermint.v1.ClientState client_state = 1;
+  // consensus_state is the consensus state of the coordinator chain.
+  ibc.lightclients.tendermint.v1.ConsensusState consensus_state = 2;
+  // initial_val_set is the initial validator set of the coordinator chain.
+  repeated .tendermint.abci.ValidatorUpdate initial_val_set = 3
+      [ (gogoproto.nullable) = false ];
+}
```
These files exist as dependencies.

MaxMustermann2 commented on x/appchain/coordinator/keeper/ibc_client.go:
```diff
@@ -0,0 +1,237 @@
+	coordinatorParams := k.GetParams(ctx)
+	clientState := coordinatorParams.TemplateClient
+	clientState.ChainId = chainID
+	// TODO(mm): Make this configurable for switchover use case
```
No, I will do this later. It is not a priority.

MaxMustermann2 committed:
doc(appchain): add more logging

MaxMustermann2 committed:
chore(lint)

MaxMustermann2 committed:
fix(ci): pin the golang lint version

MaxMustermann2 committed:
chore(lint): remove unused timestamp proto

MaxMustermann2 committed:
refactor(dogfood,operator): carve out wrapped key

The wrapped consensus key object has utility in the subscriber module as
well, and, may have some utility in the coordinator module. Hence, it is
better located in a `types` subpackage.

MaxMustermann2 committed:
Merge develop

TimmyExogenous commented on proto/exocore/appchain/common/v1/common.proto:
```diff
@@ -0,0 +1,106 @@
+
+  // Params relevant to chain operation
+  // unbonding_duration is the subscriber chain's unbonding duration. it should be less
+  // than the coordinator chain's unbonding duration.
```
Q: Why should it be less than the coordinator chain's unbonding duration?

MaxMustermann2 replied:
Hmm you're right; this is an old comment. I will remove it.

TimmyExogenous commented on x/appchain/common/types/shared_params.go:
```diff
@@ -0,0 +1,46 @@
+
+const (
+	// DefaultIBCTimeoutPeriod is the default timeout period for IBC packets.
+	DefaultIBCTimeoutPeriod = 4 * 7 * 24 * time.Hour
```
Would the IBC timeout period influence the release of unbonding assets in the coordinator chain? It seems to be greater than the unbonding duration of the coordinator chain.

MaxMustermann2 replied:
It will influence the unbonding duration, but not directly. It is the worst case timeout period. If a subscriber chain goes offline, this timeout comes into play. Otherwise, the unbonding duration is just the time to the end of the epoch + subscriber's unbonding duration + the time it takes for message delivery.

Each packet sent from the coordinator to the subscriber has this timeout period set. If the subscriber doesn't process the packet within this period, it is considered offline and removed as an AVS. During its removal, all associated unbondings are freed.

TimmyExogenous commented on x/appchain/coordinator/keeper/ibc_client.go:
```diff
@@ -0,0 +1,239 @@
+) (cctx sdk.Context, writeCache func(), err error) {
+	cctx, writeCache = ctx.CacheContext()
+	err = k.CreateClientForSubscriber(cctx, req)
+	return
```
should the `writeCache` be called before return if the err is nil?

MaxMustermann2 replied:
`writeCache` and `err` processing is handled by the caller. This is why the function has the prefix `InCachedCtx`.

TimmyExogenous commented on x/appchain/coordinator/keeper/impl_epochs_hooks.go:
```diff
@@ -0,0 +1,79 @@
+func (wrapper EpochsHooksWrapper) AfterEpochEnd(
+	ctx sdk.Context, identifier string, epoch int64,
+) {
+	// whenever an epoch ends, we should iterate over the list of pending subscriber chains
```
So we don't wait for the end of Dogfood's epoch to activate the registered app chains?

MaxMustermann2 replied:
App chains are free to choose their own epoch identifier (day, week, hour, minute). It does not have to be the same as that of x/dogfood. This is the same as any AVS.

MaxMustermann2 committed:
Merge `develop`

MaxMustermann2 committed:
fix(doc): update field comments after review

- The subscriber unbonding duration has no bounds
- The advantage of reusing an already established channel was explicitly
  listed out

bwhour commented on x/appchain/coordinator/types/msg.go:
```diff
@@ -0,0 +1,86 @@
+	}
+	// the creator is overwritten
+	r.FromAddress = creator
+	return &r
```
r.ValidateBasic() before return   ?

MaxMustermann2 replied:
Not necessary, since the Cosmos-SDK does the call before broadcast.

bwhour commented on x/appchain/coordinator/types/msg.go:
```diff
@@ -0,0 +1,86 @@
+func NewRegisterSubscriberChainRequest(creator, jsonValue string) *RegisterSubscriberChainRequest {
+	var r RegisterSubscriberChainRequest
+	if err := json.Unmarshal([]byte(jsonValue), &r); err != nil {
+		panic(fmt.Sprintf("invalid json %s", err))
```
Return errors instead of panicking, allowing the caller to handle errors ?

bwhour commented on x/appchain/coordinator/keeper/ibc_client.go:
```diff
@@ -0,0 +1,239 @@
+	}
+	operators, keys, powers = utils.SortByPower(operators, keys, powers)
+	maxVals := req.MaxValidators
+	validatorUpdates := make([]abci.ValidatorUpdate, 0, maxVals)
```
> validatorUpdates := make([]abci.ValidatorUpdate, 0, min(len(operators), int(maxVals)))  ?

bwhour commented on x/appchain/coordinator/keeper/ibc_client.go:
```diff
@@ -0,0 +1,239 @@
+		return nil, nil, err
+	}
+	operators, keys, powers = utils.SortByPower(operators, keys, powers)
+	maxVals := req.MaxValidators
```
is maxVals always > 0 ?

MaxMustermann2 replied:
Yes, this is validated in the `msg.ValidateBasic`.

bwhour commented on x/appchain/common/types/params.go:
```diff
@@ -0,0 +1,148 @@
+	downtimeJailDuration time.Duration,
+	slashFractionDoubleSign string,
+) SubscriberParams {
+	return SubscriberParams{
```
The 

> NewSubscriberParams

 function accepts a large number of parameters, and it may be considered to use the Builder pattern or the option pattern to reduce the number of parameters.

MaxMustermann2 committed:
refactor(appchain): option pattern for Subscriber

MaxMustermann2 committed:
fix(appchain-coord): remove unnecessary panic

MaxMustermann2 committed:
fix(appchain): optimize capacity of val set

MaxMustermann2 committed:
fix(appchain): option-pattern not on pointers

Since the rest of the codebase uses direct objects and not pointers,
making it pointers here would be too cumbersome.

MaxMustermann2 committed:
Merge `develop` from `upstream`

MaxMustermann2 commented on x/appchain/coordinator/keeper/ibc_client.go:
```diff
@@ -0,0 +1,248 @@
+	}
+	k.SetClientForChain(ctx, chainID, clientID)
+
+	epochInfo, _ := k.epochsKeeper.GetEpochInfo(ctx, req.EpochIdentifier)
+	initTimeoutPeriod := coordinatorParams.InitTimeoutPeriod
+	// the CurrentEpoch below is the one that has already ended, so we increment it by 1.
+	// assume we start with a value of 2 and we are giving 4 full epochs for initialization.
+	// so when epoch 6 ends, the timeout ends.
+	initTimeoutPeriod.EpochNumber += uint64(epochInfo.CurrentEpoch) + 1
```
That is unnecessary, since it is done when adding the subscriber chain.

MaxMustermann2 commented on x/appchain/coordinator/keeper/ibc_client.go:
```diff
@@ -0,0 +1,248 @@
+) (genesis commontypes.SubscriberGenesisState) {
+	store := ctx.KVStore(k.storeKey)
+	key := types.SubscriberGenesisKey(chainID)
+	k.cdc.MustUnmarshal(store.Get(key), &genesis)
+	return genesis
```
Good point, actioning this.

MaxMustermann2 commented on x/appchain/coordinator/keeper/ibc_client.go:
```diff
@@ -0,0 +1,248 @@
+	maxVals := req.MaxValidators
+	validatorUpdates := make(
+		[]abci.ValidatorUpdate, 0,
+		min(
+			// the maximum size of the validator set is the minimum of the number of operators
+			// and the max validators allowed by the subscriber chain. the capacity is not
+			// multiplied by 2 here because this is the initial validator set and there are no
+			// previous validators whose power will be set to zero.
+			int(maxVals), len(operators),
+		),
```
Built-in in `go1.21`.

MaxMustermann2 commented on utils/utils.go:
```diff
@@ -167,3 +171,42 @@ func CommonValidation[T any, V constraints.Ordered, D any](
+// the operator address (bytes, not string!) is used as a tiebreaker. The bytes
+// are preferred since that is how the operator module stores them, indexed by
+// the bytes.
+func SortByPower(
+	operatorAddrs []sdk.AccAddress,
+	pubKeys []types.WrappedConsKey,
+	powers []int64,
+) ([]sdk.AccAddress, []types.WrappedConsKey, []int64) {
+	// Create a slice of indices
+	indices := make([]int, len(powers))
+	for i := range indices {
+		indices[i] = i
+	}
+
+	// Sort the indices slice based on the powers slice
+	// Since the operator address is unique, SliceStable is not needed
+	sort.Slice(indices, func(i, j int) bool {
+		if powers[indices[i]] == powers[indices[j]] {
+			// ascending order of operator address as tiebreaker
+			return bytes.Compare(operatorAddrs[indices[i]], operatorAddrs[indices[j]]) < 0
+		}
+		// descending order of power
+		return powers[indices[i]] > powers[indices[j]]
+	})
+
+	// Reorder all slices using the sorted indices
+	sortedOperatorAddrs := make([]sdk.AccAddress, len(operatorAddrs))
+	sortedPubKeys := make([]types.WrappedConsKey, len(pubKeys))
+	sortedPowers := make([]int64, len(powers))
+	for i, idx := range indices {
+		sortedOperatorAddrs[i] = operatorAddrs[idx]
+		sortedPubKeys[i] = pubKeys[idx]
+		sortedPowers[i] = powers[idx]
+	}
+	return sortedOperatorAddrs, sortedPubKeys, sortedPowers
+}
```
The reason these are separate slices is because the power is queried using a different logic.

MaxMustermann2 commented on utils/utils.go:
```diff
@@ -167,3 +171,42 @@
+// the operator address (bytes, not string!) is used as a tiebreaker. The bytes
+// are preferred since that is how the operator module stores them, indexed by
+// the bytes.
+func SortByPower(
+	operatorAddrs []sdk.AccAddress,
+	pubKeys []types.WrappedConsKey,
+	powers []int64,
+) ([]sdk.AccAddress, []types.WrappedConsKey, []int64) {
```
Added comment `The caller must ensure that the slices are of the same length.`

MaxMustermann2 commented on x/dogfood/keeper/validators.go:
```diff
@@ -36,25 +38,12 @@ func (k Keeper) UnbondingTime(ctx sdk.Context) time.Duration {
-		// the address is just derived from the public key and
-		// has no correlation with the operator address on Exocore.
-		addr := sdk.GetConsAddress(pubkey)
+		addr := change.Key.ToConsAddr()
```
This is already enforced by the callers everywhere, added a comment.

MaxMustermann2 commented on x/dogfood/keeper/validators.go:
```diff
@@ -99,7 +88,7 @@
 			if change.Power > 0 {
 				// create a new validator.
-				ocVal, err := types.NewExocoreValidator(addr, change.Power, pubkey)
+				ocVal, err := types.NewExocoreValidator(addr, change.Power, change.Key.ToSdkKey())
```
This is already enforced by the callers everywhere, added a comment.

MaxMustermann2 commented on x/dogfood/keeper/validators.go:
```diff
@@ -122,7 +111,10 @@
 			}
 		}
-		ret = append(ret, change)
+		ret = append(ret, abci.ValidatorUpdate{
+			PubKey: *change.Key.ToTmProtoKey(),
+			Power:  change.Power,
+		})
```
This can never be nil as long as `change.Key` is not nil, which is something the caller must ensure.

MaxMustermann2 committed:
fix: respond to AI comments

MaxMustermann2 committed:
fix(coordinator): correct epochs calc

bwhour commented on x/appchain/common/types/params.go:
```diff
@@ -0,0 +1,245 @@
+		WithBlocksPerDistributionTransmission(blocksPerDistributionTransmission).
+		WithSubscriberRedistributionFraction(subscriberRedistributionFraction).
+		WithRewardDenom(rewardDenom).
+		WithIBCTimeoutPeriod(ibcTimeoutPeriod).
```
Do we need  add range checks for the time.Duration field to prevent extreme values ?

MaxMustermann2 replied:
The assumption is that these values are set by the subscriber appchain. If they set unrealistic values, the operators will not opt-in to the chain, given it's a free market for staking. We can only validate that the values aren't zero or something, which we have done in `ValidateBasic`

MaxMustermann2 committed:
Merge branch 'develop' into appchains

bwhour commented on x/appchain/coordinator/keeper/ibc_client.go:
```diff
@@ -0,0 +1,251 @@
+		ctx, operators, chainIDWithoutRevision,
+	)
+	if err != nil {
+		k.Logger(ctx).Error("error getting vote power for chain", "error", err)
```
is the _**err**_ include "chainIDWithoutRevision" info ?

MaxMustermann2 replied:
Yes; added a comment to this effect just above.

bwhour commented on x/appchain/coordinator/keeper/impl_epochs_hooks.go:
```diff
@@ -0,0 +1,79 @@
+			// no re-attempts will be made for this subscriber
+			ctx.Logger().Error(
+				"subscriber client not created",
+				"chainID", subscriber,
```
subscriber.ChainID ?

bwhour reviewed (approved):
The whole logic looks good.

MaxMustermann2 committed:
fix: build and tests

- The `github.com/ExocoreNetwork/exocore/types` package is a bit finicky
  to deal with, and hence, the wrapped key has been moved to its own
  subpackage within that.
- Some files that were copied word-to-word in our codebase were never
  being used but instead referred from the parent codebase. This
  resulted in test errors about duplicate `proto` registration. Hence,
  these files have been removed.

MaxMustermann2 committed:
fix(appchain): respond to review comments

MaxMustermann2 commented:
This PR is ready and may be merged after the testnet v6 release is cut.
